--provider global flag for kompose

surajssd opened the pull request:
Now a user can select a provider using global flag `--provider=openshift` to select openshift provider
or `--provider-kubernetes` to select kubernetes provider if nothing is provided kubernetes is the default provider.

Fixes #179

kadel commented:
Hmm, today on train to Berlin I've worked on same thing as part of work on `kompose up` for openshift.

I have slightly different approach. I wanted to avoid manual validation as much as possible and leave it on cli library. I've used `app.Before` to add Commands depending on specified provider.
You can see my wip in branch openshift-up in my fork https://github.com/skippbox/kompose/compare/master...kadel:openshift-up

kadel commented:
Today on train to Berlin I've worked on same thing as part of work on `kompose up` for openshift. :-)

Because I wanted `--help` to show only options that are valid for given provider, I've done this slightly differently.

I've used `app.Before` to add Commands depending on specified provider.
You can see my wip in branch openshift-up in my fork https://github.com/skippbox/kompose/compare/master...kadel:openshift-up 

I don't know how much time I'm going to have this week to clean it up and finish. But If you want to have provider flag sooner, look at my branch and you can use it ;-)

surajssd commented:
@kadel thanks for pointers, this helps with the missing point I had!

surajssd commented:
@kadel addressed the issues you mentioned!

Now output looks like this:

for openshift
```bash
$ kompose --provider=openshift convert -h
NAME:
   kompose convert - Convert Docker Compose file (e.g. docker-compose.yml) to Kubernetes objects

USAGE:
   kompose convert [command options] [arguments...]

OPTIONS:
   --out, -o                    Specify file name in order to save objects into [$OUTPUT_FILE]
   --replicas "1"               Specify the number of replicas in the generated resource spec (default 1)
   --chart, -c                  Create a chart deployment
   --yaml, -y                   Generate resource file in yaml format
   --stdout                     Print converted objects to stdout
   --deploymentconfig, --dc     Generate a DeploymentConfig for OpenShift
```

for k8s

```bash
$ kompose convert -h
NAME:
   kompose convert - Convert Docker Compose file (e.g. docker-compose.yml) to Kubernetes objects

USAGE:
   kompose convert [command options] [arguments...]

OPTIONS:
   --out, -o                            Specify file name in order to save objects into [$OUTPUT_FILE]
   --replicas "1"                       Specify the number of replicas in the generated resource spec (default 1)
   --chart, -c                          Create a chart deployment
   --yaml, -y                           Generate resource file in yaml format
   --stdout                             Print converted objects to stdout
   --deployment, -d                     Generate a deployment resource file (default on)
   --daemonset, --ds                    Generate a daemonset resource file
   --replicationcontroller, --rc        Generate a replication controller resource file

```

kadel commented on cli/app/app.go:
```diff
@@ -39,130 +40,133 @@ import (
+		if !opt.CreateD && !opt.CreateDS && !opt.CreateRC {
+			opt.CreateD = true
+		}
+		if opt.CreateDeploymentConfig {
```
this check is not needed,  this can never happen

surajssd replied:
Oh yes, now this is not needed since command line handling will take care of it :)

kadel commented on cli/app/app.go:
```diff
@@ -39,130 +40,133 @@ import (
+		if !opt.CreateDeploymentConfig {
+			opt.CreateDeploymentConfig = true
+		}
 		if opt.CreateD {
```
this check is not needed, this can never happen same for CreateDS and CreateRS

kadel commented on cli/command/command.go:
```diff
@@ -18,11 +18,72 @@ package command
+	provider := strings.ToLower(c.GlobalString("provider"))
+	switch provider {
+	case "kubernetes":
+		k8sControllers := []cli.Flag{
```
can we create whole new `cli.Command` here?  So we can set usage help for every provider independently?
Or set `commands.Usage` here

If I run `kompose --provider openshift convert -h ` It is still displaying usage mentioning Kubernetes

```
NAME:
   kompose convert - Convert Docker Compose file (e.g. docker-compose.yml) to Kubernetes objects
....
```

This is also relying on `ConvertCommand` always being  first in that slice :-( This might lead to errors in the future. At least there should be comment in `app.Commands`

surajssd replied:
Yes sure we can create new `cli.Command`, and nice catch of that usage string BTW!

kadel commented on cli/main/main.go:
```diff
@@ -33,7 +32,7 @@ func main() {
+	app.Before = command.BeforeApp
 	app.Flags = append(command.CommonFlags())
 	app.Commands = []cli.Command{
 		command.ConvertCommand(),
```
in `command.BeforeApp` you are counting on ConvertCommand being always in the first place here.
If you want it like that, here should be comment explaining that this cannot be moved.

surajssd replied:
I stuck to the way you did it, add the command itself in `BeforeApp` function, so this is not needed I think.

surajssd commented:
Changed it now to have new `cli.Command` altogether, so usage is changed, added new function that returns common flags.

```bash
$ kompose --provider=openshift convert -h
NAME:
    - Convert Docker Compose file (e.g. docker-compose.yml) to OpenShift objects

USAGE:
    [command options] [arguments...]

OPTIONS:
   --deploymentconfig, --dc     Generate a OpenShift DeploymentConfig object
   --out, -o                    Specify file name in order to save objects into [$OUTPUT_FILE]
   --replicas "1"               Specify the number of replicas in the generated resource spec (default 1)
   --yaml, -y                   Generate resource file in yaml format
   --stdout                     Print converted objects to stdout
```

kadel commented on cli/command/command.go:
```diff
@@ -18,60 +18,121 @@ package command
+	return command
+}
+
+func commonConvertFlags() []cli.Flag {
```
:+1:

surajssd commented:
@kadel sure will wait for @sebgoa or @ngtuna to try it out once!

ngtuna commented:
Great work @surajssd  :+1: I will try it today. Thanks

ngtuna commented:
Hey @surajssd, at first test I see `$ kompose --help` lists out `up` and `down` command only. Convert is missing.

```console
$ kompose --help
...
COMMANDS:
    up		Deploy your Dockerized application to Kubernetes (default: creating Kubernetes deployment and service)
    down	Delete instantiated services/deployments from kubernetes
...
```

ngtuna commented:
Otherwise LGTM. I like the idea of changing usage between providers :100:

ngtuna commented:
Then provider owners have their own space to arrange outputs.

surajssd commented:
@ngtuna nice catch about convert not being visible. Made changes accordingly here https://github.com/skippbox/kompose/pull/182/files#diff-3145698b159fb0a026984b4a017a8b24R67 also added note here https://github.com/skippbox/kompose/pull/182/files#diff-d9814220427d68b7f7ade416b1440032R38

surajssd commented:
And also it shows up under help

```bash
$ kompose -h
NAME:
[SNIP]
   Skippbox Kompose Contributors <https://github.com/skippbox/kompose> 
   
COMMANDS:
    convert     Convert Docker Compose file (e.g. docker-compose.yml) to Kubernetes/OpenShift objects
    up          Deploy your Dockerized application to Kubernetes (default: creating Kubernetes deployment and service)
    down        Delete instantiated services/deployments from kubernetes

GLOBAL OPTIONS:
[SNIP]
```

ngtuna commented:
Thanks @surajssd . Merging now for the release tonight.

surajssd commented:
Thanks @ngtuna !
